### PR TITLE
Check for IO errors when loading binary shader

### LINKED ISSUE #740: Cached shader loader keeps reading garbage
For some reason the cached GL program binary loader keeps reading garbage from the cache file. I have added some checks to stop this crashing the app, but I need to figure out why it is reading garbage in the first place.

The relevant code is fairly simple C++ iostreams code so I don't see why it is going crazy. See https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/shader/shader.cpp#L28-L37

### PR DESCRIPTION
**Pull request by ljbade**
This fixes #740 

I had to add checks for IO errors (I thought iostream threw on errors).

I also put in the sanity check code from android-mason for good measure.

### CODE REVIEW

**Comment by ljbade** [comment]
I just found out that you can actually enable C++ exceptions with http://en.cppreference.com/w/cpp/io/basic_ios/exceptions

@jfirebaugh Would this be a better way to handle errors?

**Comment by kkaefer** [comment]
Can we use our caching infrastructure instead of writing directly to disk? E.g., we could introduce a `cache://` scheme that only reads/writes to the cache, but doesn't attempt to load anything from disk. If it can't find the URL in cache, it just returns a not found error and the requesting application can generate the resource manually and then store it into the cache.

**Comment by ljbade** [comment]
I suppose we could do that. Does the caching infrastructure support overwriting?

Since binary shader depends on GL state combinations, loading different maps/areas could trigger more information to be included in the binary shader each time the app is closed.

**Comment by ljbade** [comment]
@kkaefer Perhaps I should cut a new ticket for that? In the mean time this pull request will at least fix a crash that was holding me up on working on another bug.

**Comment by jfirebaugh** [comment]
Yeah, I generally prefer exceptions to manual error checking. In this case, I think the code should also handle any other exceptions (e.g. from `MBGL_CHECK_ERROR`) by abandoning the cached shaders, so it's probably easiest if the whole block is surrounded with `try { } catch (...) { }`.

**Comment by ljbade** [comment]
@jfirebaugh converted to exceptions

Also fixed the missing call to map::terminate which was causing havok with shader destructor. I think this might have been causing some other crashes on Android too.

Can you review and merge if all is OK?

**Comment by ljbade** [comment]
Also fix rare crash when HTTP request finishes after Map termination begins.

File: src/mbgl/shader/shader.cpp [hunk 1]
```diff
@@ -26,26 +26,53 @@ Shader::Shader(const char *name_, const GLchar *vertSource, const GLchar *fragSo
...
+            MBGL_CHECK_ERROR(glGetProgramiv(program, GL_LINK_STATUS, &status));
+            if (status == GL_TRUE) {
+                skipCompile = true;
+            }
+
+        } catch(std::exception e) {
+            Log::Error(Event::Shader, "Loading binary shader failed!");
+
+            // Delete the bad file
+            std::remove(binaryFileName.c_str());
```
Comment (jfirebaugh) [inline src/mbgl/shader/shader.cpp]: Which [`std::remove`](http://en.cppreference.com/w/cpp/algorithm/remove) is this? I'm surprised a one-argument call compiles.

**Pull request merged**